pdl_syntax.md

```diff
@@ -0,0 +1,828 @@
+For example : 
+
+```
+namespace com.linkedin.pegasus.generator.test.idl.escaping
+
+record PropertyKeyEscaping {
+  @`namespace` = "foo.bar"
+  @`test.path` = 1
+  @validate.`com.linkedin.CustomValidator` = "foo"
+  aField: string
+}
+```
+
+
+## Primitive Types
+
+The Pegasus primitive types are : int, long, float, double, boolean, string and bytes.
+
+For Example :
+```
+namespace com.linkedin.pegasus.generator.test.idl.records
+
+record WithPrimitives {
+  intField: int
+  longField: long
+  floatField: float
+  doubleField: double
+  booleanField: boolean
+  stringField: string
+  bytesField: bytes
+}
+```
+
+## Array Type
+
+Pegesus Arrays are defined within a items type.
+
+For Example : 
+```
+namespace com.linkedin.pegasus.generator.test.idl.arrays
+
+record WithPrimitivesArray {
+  ints: array[int]
+  longs: array[long]
+  floats: array[float]
+  doubles: array[double]
+  booleans: array[boolean]
+  strings: array[string]
+  bytes: array[bytes]
+}
+```
+
+Record Arrays :
+
+```
+namespace com.linkedin.pegasus.generator.test.idl.arrays
+
+import com.linkedin.pegasus.generator.test.idl.enums.Fruits
+import com.linkedin.pegasus.generator.test.idl.records.Empty
+
+record WithRecordArray {
+  empties: array[Empty]
+  fruits: array[Fruits]
+}
+```
+
+
+## Map Type
+
+Pegasus maps are defined with a values type and an optional key type.
+
+For example : 
+```
+namespace com.linkedin.pegasus.generator.test.idl.maps
+
+record WithPrimitivesMap {
+  ints: map[string, int]
+  longs: map[string, long]
+  floats: map[string, float]
+  doubles: map[string, double]
+  booleans: map[string, boolean]
+  strings: map[string, string]
+  bytes: map[string, bytes]
+}
+```
+Custom Types Map :
+```
+namespace com.linkedin.pegasus.generator.test.idl.maps
+
+import com.linkedin.pegasus.generator.test.idl.customtypes.CustomInt
+
+record WithCustomTypesMap {
+  ints: map[string, CustomInt]
+}
+```
+
+Complex Types Map :
+```
+namespace com.linkedin.pegasus.generator.test.idl.maps
+
+import com.linkedin.pegasus.generator.test.idl.enums.Fruits
+import com.linkedin.pegasus.generator.test.idl.records.Empty
+import com.linkedin.pegasus.generator.test.idl.records.Simple
+import com.linkedin.pegasus.generator.test.idl.`fixed`.Fixed8
+
+record WithComplexTypesMap {
+  empties: map[string, Empty]
+  fruits: map[string, Fruits]
+  arrays: map[string, array[Simple]]
+  maps: map[string, map[string, Simple]]
+  unions: map[
+    string,
+    typeref WithComplexTypesMapUnion = union[int, string, Simple]
+  ]
+  `fixed`: map[string, Fixed8]
+}
+```
+
+**Note**: 
+The key must always be "string".
+
+
+## Union Type
+
+A union type may be defined with any number of member types. Each member may be any pegasus type except union type. 
+Each member can be primitive, record, enum map or array type.
+
+For example :
+```
+namespace com.linkedin.pegasus.generator.test.idl.unions
+
+record WithPrimitivesUnion {
+  `union`: union[int, long, float, double, boolean, string, bytes]
```

Comment:
Suggest picking a different name for the field to avoid confusion

Reply:
Updated the name.